add self-assessment.md and joint-review.md (#81)
kubeedge/community: sig-security/joint-review.md

### Attack Risks and Effects

- Data saved in one database on the edge node, if an attacker attacks an edge database, it will cause denial of service in off-line scenarios.
- External App can request Kubernetes resource to Meta-Server through list-watch method, then the request will be passed to CloudCore and finally to the api-server of the cluster. If an app was exploited by attacker, a DOS attack to list-watch server can be launched.

### Security Degradation

Exploiting vulnerabilities in management on the cloud can lead to maliciously deleting resources or bringing the whole cluster down. If an attacker intercepts tokens used for communicating with CloudCore, he can register a fake node and render the CloudCore unusable. If attackers gain access to the EdgeCore database, he could manipulate critical metadata such as adding or deleting existing pods or change the host mode or image pull address of pods.

### Compensating Mechanisms

> - Please see the mitigations described at [here](https://github.com/kubeedge/community/blob/master/sig-security/sig-security-audit/KubeEdge-threat-model-and-security-protection-analysis.md#kubeedge-threat-model-and-mitigations), under the section `KubeEdge threat-model and mitigations`.

Reviewer: It's recommended to include brief summary of the "threat-model and mitigations" here, e.g. number of risks and mitigations provided.

Reviewer: Also, instead of `[here]`, I'd suggest to use the document name

## Threat Model

Please see the detailed threat model analysis paper of KubeEdge at [here](https://github.com/kubeedge/community/blob/master/sig-security/sig-security-audit/KubeEdge-threat-model-and-security-protection-analysis.md).

### Identity Theft

> Please refer to the above section `Threat Model`.

Reviewer: Please add quote from the "threat model"

### Compromise

> Please refer to the above section `Threat Model`.

Reviewer: same as above

### Denial of Service

Please refer to the above section `Threat Model`.

## Secure Development Practices

KubeEdge has achieved the passing level criteria for [CII Best Practices](https://bestpractices.coreinfrastructure.org/en/projects/3018).

- Development Pipeline

> All code is maintained in [Git](https://github.com/kubeedge/kubeedge) and changes must be reviewed by maintainers and must pass all unit and e2e tests. Code changes are submitted via Pull Requests (PRs) and must be signed. Commits to the `main` branch directly are not allowed.

Reviewer: becides UT and e2e, we have static checks, verifications on gofmt, go lint, go vet, vendors, and fossa check.

- Communication Channels

>   - Internal. How do team members communicate with each other?
>     Team members communicate with each other frequently through [Slack Channel](https://join.slack.com/t/kubeedge/shared_invite/enQtNjc0MTg2NTg2MTk0LWJmOTBmOGRkZWNhMTVkNGU1ZjkwNDY4MTY4YTAwNDAyMjRkMjdlMjIzYmMxODY1NGZjYzc4MWM5YmIxZjU1ZDI), [KubeEdge sync meeting](https://zoom.us/my/kubeedge), and team members will new a [issue](https://github.com/kubeedge/kubeedge/issues) to make a deep discussion if necessary.

Reviewer: ```suggestion
  - Internal. How do team members communicate with each other?
  
    Team members communicate with each other frequently through [Slack Channel](https://join.slack.com/t/kubeedge/shared_invite/enQtNjc0MTg2NTg2MTk0LWJmOTBmOGRkZWNhMTVkNGU1ZjkwNDY4MTY4YTAwNDAyMjRkMjdlMjIzYmMxODY1NGZjYzc4MWM5YmIxZjU1ZDI), [KubeEdge sync meeting](https://zoom.us/my/kubeedge), and team members will new a [issue](https://github.com/kubeedge/kubeedge/issues) to make a deep discussion if necessary.

```

>   - Inbound. How do users or prospective users communicate with the team?
>     Users or prospective users usually communicate with the team through [Slack Channel](https://kubeedge.slack.com/archives/CDXVBS085), you can new a [issue](https://github.com/kubeedge/kubeedge/issues) to get further help from the team, and [KubeEdge mailing list](https://groups.google.com/forum/#!forum/kubeedge) is also available. Besides, we have regular [community meeting](https://zoom.us/my/kubeedge) (includes SIG meetings) alternative between Europe friendly time and Pacific friendly time, all these meetings are publicly accessible and meeting records are uploaded to YouTube.

Reviewer: ```suggestion
  - Inbound. How do users or prospective users communicate with the team?
  
    Users or prospective users usually communicate with the team through [Slack Channel](https://kubeedge.slack.com/archives/CDXVBS085), you can new a [issue](https://github.com/kubeedge/kubeedge/issues) to get further help from the team, and [KubeEdge mailing list](https://groups.google.com/forum/#!forum/kubeedge) is also available. Besides, we have regular [community meeting](https://zoom.us/my/kubeedge) (includes SIG meetings) alternative between Europe friendly time and Pacific friendly time, all these meetings are publicly accessible and meeting records are uploaded to YouTube.

```

    Regular Community Meeting:

    - Europe Time: **Wednesdays at 16:30-17:30 Beijing Time** (biweekly, starting from Feb. 19th 2020). ([Convert to your timezone.](https://www.thetimezoneconverter.com/?t=16%3A30&tz=GMT%2B8&))
    - Pacific Time: **Wednesdays at 10:00-11:00 Beijing Time** (biweekly, starting from Feb. 26th 2020). ([Convert to your timezone.](https://www.thetimezoneconverter.com/?t=10%3A00&tz=GMT%2B8&))

> - Outbound. How do you communicate with your users? (e.g. flibble-announce@ mailing list)
>   KubeEdge communicate with users through [Slack Channel](https://kubeedge.slack.com/archives/CUABZBD55), [issue](https://github.com/kubeedge/kubeedge/issues), [KubeEdge sync meeting](https://zoom.us/my/kubeedge), [KubeEdge mailing list](https://groups.google.com/forum/#!forum/kubeedge), as for security issues, we provide the communication channels as follows.

Reviewer: ```suggestion
- Outbound. How do you communicate with your users? (e.g. flibble-announce@ mailing list)

  KubeEdge communicate with users through [Slack Channel](https://kubeedge.slack.com/archives/CUABZBD55), [issue](https://github.com/kubeedge/kubeedge/issues), [KubeEdge sync meeting](https://zoom.us/my/kubeedge), [KubeEdge mailing list](https://groups.google.com/forum/#!forum/kubeedge), as for security issues, we provide the communication channels as follows.

```

>   - Security email group
>     You can email [kubeedge-security](mailto:cncf-kubeedge-security@lists.cncf.io) to report a vulnerability, and the KubeEdge security team will disclose to the distributors through [distributors announce list](mailto:cncf-kubeedge-distrib-announce@lists.cncf.io), see more details [here](https://github.com/kubeedge/kubeedge/security/policy).

Reviewer: ```suggestion
  - Security email group
  
    You can email [kubeedge-security](mailto:cncf-kubeedge-security@lists.cncf.io) to report a vulnerability, and the KubeEdge security team will disclose to the distributors through [distributors announce list](mailto:cncf-kubeedge-distrib-announce@lists.cncf.io), see more details [here](https://github.com/kubeedge/kubeedge/security/policy).

```

## Security Issue Resolution

### Responsible Disclosures Process

KubeEdge project vulnerability handling related processes are recorded in [Security Policy](https://github.com/kubeedge/kubeedge/security/policy), Related security vulnerabilities can be reported and communicated via email `cncf-kubeedge-security@lists.cncf.io`.

### Incident Response

See the [KubeEdge releases page](https://github.com/kubeedge/kubeedge/releases) for information on supported versions of KubeEdge. Once the fix is confirmed, the Security Team will patch the vulnerability in the next patch or minor release, and backport a patch release into the latest three minor releases.

The release of low to medium severity bug fixes will include the fix details in the patch release notes. Any public announcements sent for these fixes will be linked to the release notes.

### Closed security issues and vulnerabilities

> According to the vulnerability release process, the KubeEdge team published the vulnerability in Security Advisories on Github. All the vulnerabilities have applied for CVE ID, mainly focusing on the problems of memory exhaustion and access exceptions. For details, please see: https://github.com/kubeedge/kubeedge/security/advisories

Reviewer: ```suggestion
According to the vulnerability release process, the KubeEdge team published the vulnerability in Security Advisories on Github. All the vulnerabilities have applied for CVE ID, mainly focusing on the problems of memory exhaustion and access exceptions. For details, please see: https://github.com/kubeedge/kubeedge/security/advisories?state=published

```

### Hands-on review result

General comments and summary of the hands-on review with any recommendations worth
 noting.  If nothing found use the below example:

> TAG-Security's hands-on review did not reveal any significant or notable security findings for [project]. This outcome does not indicate that none exist, rather that none were discovered.

## Roadmap

* **Project Next Steps.** *Link to your general roadmap, if available, then list prioritized next steps that may have an impact on the risk profile of your project, including anything that was identified as part of this review.*
* **CNCF Requests.** *In the initial draft, please include whatever you believe the CNCF could assist with that would increase security of the ecosystem.*
  - Assess KubeEdge CI workflows and provide security hardening methods.
  - According to the threat model, provide a way to harden security from the overall perspective.

## Appendix

### Known Issues Over Time.

For details, please see: 

https://github.com/kubeedge/kubeedge/security/advisories

### Case Studies

KubeEdge has been widely adopted in industries including transportation, energy, Internet, CDN, manufacturing, smart campus etc.

> * Xinghai IoT is an IoT company that provides comprehensive smart building solutions by leveraging a construction IoT platform, intelligent hardware, and AI. Xinghai IoT built a smart campus with cloud-edge-device synergy based on KubeEdge and its own Xinghai IoT cloud platform, greatly improving the efficiency of campus management. With AI assistance, nearly 30% of the repetitive work is automated. In the future, Xinghai IoT will continue to collaborate with KubeEdge to launch KubeEdge-based smart campus solutions.

Reviewer: Same as above, include more according to our latest updates

* In China’s highway electronic toll collection (ETC) system, KubeEdge helps manage nearly 100,000 edge nodes and more than 500,000 edge applications in 29 of China’s 34 provinces, cities and autonomous regions. With these applications, the system processes more than 300 million data records daily and has improved traffic efficiency at the toll stations by 10 times.
* China Mobile On-line Marketing Service Center, is a secondary organ of the China Mobile Communications Group, holds the world’s largest call center with 44,000 agents, 900 million users, and 53 customer service centers, builds a cloud-edge synergy architecture consisting of two centers and multiple edges based on KubeEdge. See more details [here](https://www.cncf.io/blog/2021/08/16/china-mobile-kubeedge-based-customer-service-platform-featuring-edge-cloud-synergy).
* e-Cloud of China Telecom, uses KubeEdge to manage CDN edge nodes, automatically deploy and upgrade CDN edge services, and implement edge service disaster recovery (DR) when it migrates its CDN services to the cloud. See more details [here](https://www.cncf.io/blog/2022/03/18/e-cloud-large-scale-cdn-using-kubeedge).

### Related Projects / Vendors

> Another edge computing that relative to Kubernetes is OpenYurt, it uses a centralized Kubernetes control plane residing in the cloud site to manage multiple edge nodes residing in the edge sites. OpenYurt manage edge nodes through native Kubelet component, each edge node has moderate compute resources available in order to run edge applications plus the required OpenYurt components. It provides a edge node daemon YurtHub that serves as a proxy for the outbound traffic from typical Kubernetes node daemons such as Kubelet, Kubeproxy, CNI plugins, etc.

Reviewer: This part we can describe more about related projects those are working on edge computing scenarios and can be leveraged by KubeEdge to enforce better user/developer experience when building cloud-edge platforms or systems.

### Expected Attacker Capabilities

All internal components of KubeEdge use TLS to communicate with each other. There are currently no known vulnerabilities and operators have the ability to exploit TLS for the communications between internal components. As for the components (ServiceBus of EdgeCore) that communicate with external applications, TLS is not enforced, but the service is only exposed to the local host and the KubeEdge team has documented the risks of an insecure setup and made it harder for operators to deploy insecurely.

Areas that are potentially susceptible to attacks would require credential handling, for example, token-based single sign-on from edge to cloud. However, all of these use some forms of tokens based authentication and we assume the attacker cannot easily break existing cryptographic algorithms like AES or SHA256.

### Attack Risks and Effects

- For data saved in one database on the edge node, if an attacker attacks an edge database, it will cause denial of service in off line scenarios.

> - External apps can request Kubernetes resources to Meta-Server through list-watch, then the request will be passed to CloudCore and finally to the api-server of the cluster. If an app was exploited by an attacker, a DOS attack to list-watch server can be launched.

Reviewer: ```suggestion
- External apps can request Kubernetes resources to MetaServer through list-watch, then the request will be passed to CloudCore and finally to the api-server of the cluster. If an app was exploited by an attacker, a DOS attack to list-watch server can be launched.

```

### Security Degradation

Exploiting vulnerabilities in management on the cloud can lead to maliciously deleting resources or bringing the whole cluster down. If an attacker intercepts tokens used for communicating with CloudCore, he can register fake nodes and render the CloudCore unusable. If attackers gain access to the EdgeCore database, he could manipulate critical metadata such as adding or deleting existing pods or change the host mode or image pull address of pods.

### Compensating Mechanisms

According to the trust boundary, we list 17 corresponding mitigations currently available in the KubeEdge project and 11 security reinforce recommendations for users and developers. Please see the details at [KubeEdge threat-model and mitigations](https://github.com/kubeedge/community/blob/master/sig-security/sig-security-audit/KubeEdge-threat-model-and-security-protection-analysis.md#kubeedge-threat-model-and-mitigations).

## Threat Model

> KubeEdge security team did an overall system security analysis of KubeEdge, mainly based on the ASTRIDE Low Level threat modeling method and the [KubeEdge security audit](https://github.com/kubeedge/community/blob/master/sig-security/sig-security-audit/KubeEdge-security-audit-2022.pdf). 

Reviewer: ```suggestion
KubeEdge security team did an [overall system security analysis of KubeEdge](https://github.com/kubeedge/community/blob/master/sig-security/sig-security-audit/KubeEdge-threat-model-and-security-protection-analysis.md#kubeEdge-threat-model-and-mitigations), mainly based on the [STRIDE threat modeling](https://en.wikipedia.org/wiki/STRIDE_(security)) and the [KubeEdge security audit](https://github.com/kubeedge/community/blob/master/sig-security/sig-security-audit/KubeEdge-security-audit-2022.pdf). 
```

* China Mobile On-line Marketing Service Center, a secondary organ of the China Mobile Communications Group, which holds the world’s largest call center with 44,000 agents, 900 million users, and 53 customer service centers, builds a cloud-edge synergy architecture consisting of two centers and multiple edges based on KubeEdge. See more details [here](https://www.cncf.io/blog/2021/08/16/china-mobile-kubeedge-based-customer-service-platform-featuring-edge-cloud-synergy).
* Xinghai IoT is an IoT company that provides comprehensive smart building solutions by leveraging a construction IoT platform, intelligent hardware, and AI. Xinghai IoT built a smart campus with cloud-edge-device synergy based on KubeEdge and its own Xinghai IoT cloud platform, greatly improving the efficiency of campus management. With AI assistance, nearly 30% of the repetitive work is automated. In the future, Xinghai IoT will continue to collaborate with KubeEdge to launch KubeEdge-based smart campus solutions.

> * SF, the largest integrated logistics service provider in China and the fourth largest in the world, has completely transformed their supply chains. Based on KubeEdge, SF Express smart supply chain system realizes edge-cloud synergy, edge data collection, real-time edge data processing, massive IoT device management, and creates a more stable, secure, and reliable edge industrial IoT system.

Reviewer: ```suggestion
* [SF Express](https://www.sf-express.com/), the largest integrated logistics service provider in China and the fourth largest in the world, has completely transformed their supply chains. Based on KubeEdge, SF Express smart supply chain system realizes edge-cloud synergy, edge data collection, real-time edge data processing, massive IoT device management, and creates a more stable, secure, and reliable edge industrial IoT system.

```
